[Gradient Compression] Refactor tensor grouping in PowerSGD

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53099 [Gradient Compression] Make GradBucket class public
* **#52981 [Gradient Compression] Refactor tensor grouping in PowerSGD**

No need to create a hard boundary between rank-1 tensors and high-rank tensors, since some high-rank tensors will not be compressed if the compression cannot save enough bandwidth, according to `_should_compress` function.

Therefore, refactor and simplify the tensor grouping logic, which addresses the comment in https://github.com/pytorch/pytorch/pull/52541#discussion_r580867311

Differential Revision: [D26713689](https://our.internmc.facebook.com/intern/diff/D26713689/)

Commit:
[Gradient Compression] Refactor tensor grouping in PowerSGD

No need to create a hard boundary between rank-1 tensors and high-rank tensors, since some high-rank tensors will not be compressed if the compression cannot save enough bandwidth, according to `_should_compress` function.

Therefore, refactor and simplify the tensor grouping logic, which addresses the comment in https://github.com/pytorch/pytorch/pull/52541#discussion_r580867311

Differential Revision: [D26713689](https://our.internmc.facebook.com/intern/diff/D26713689/)

[ghstack-poisoned]

SciPioneer commented:
@tvogels Can you take a look?

Particularly, I used `n, m = tensor.shape[0], tensor.numel() / tensor.shape[0]` instead of
`n, m = tensor.shape[0], tensor.nelement() // tensor.shape[0]` in your [comment](https://github.com/pytorch/pytorch/pull/52541#discussion_r580867311).

Commit:
Update on "[Gradient Compression] Refactor tensor grouping in PowerSGD"


No need to create a hard boundary between rank-1 tensors and high-rank tensors, since some high-rank tensors will not be compressed if the compression cannot save enough bandwidth, according to `_should_compress` function.

Therefore, refactor and simplify the tensor grouping logic, which addresses the comment in https://github.com/pytorch/pytorch/pull/52541#discussion_r580867311

Differential Revision: [D26713689](https://our.internmc.facebook.com/intern/diff/D26713689/)

[ghstack-poisoned]

tvogels commented:
Looks good to me.

Commit:
Update on "[Gradient Compression] Refactor tensor grouping in PowerSGD"


No need to create a hard boundary between rank-1 tensors and high-rank tensors, since some high-rank tensors will not be compressed if the compression cannot save enough bandwidth, according to `_should_compress` function.

Therefore, refactor and simplify the tensor grouping logic, which addresses the comment in https://github.com/pytorch/pytorch/pull/52541#discussion_r580867311

Differential Revision: [D26713689](https://our.internmc.facebook.com/intern/diff/D26713689/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor tensor grouping in PowerSGD"


No need to create a hard boundary between rank-1 tensors and high-rank tensors, since some high-rank tensors will not be compressed if the compression cannot save enough bandwidth, according to `_should_compress` function.

Therefore, refactor and simplify the tensor grouping logic, which addresses the comment in https://github.com/pytorch/pytorch/pull/52541#discussion_r580867311

Differential Revision: [D26713689](https://our.internmc.facebook.com/intern/diff/D26713689/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor tensor grouping in PowerSGD"


No need to create a hard boundary between rank-1 tensors and high-rank tensors, since some high-rank tensors will not be compressed if the compression cannot save enough bandwidth, according to `_should_compress` function.

Therefore, refactor and simplify the tensor grouping logic, which addresses the comment in https://github.com/pytorch/pytorch/pull/52541#discussion_r580867311

Differential Revision: [D26713689](https://our.internmc.facebook.com/intern/diff/D26713689/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -180,36 +180,34 @@ def powerSGD_hook(
+    1. Views the input flattened 1D gradient tensor as a list of per-parameter tensors, and divides all the tensors into two groups:
 
-    2. Divides all the tensors into two groups:
+        1.1 The tensors that should be compressed before allreduce, because the compression can give enough saving in bandwidth.
```
Do we specify somewhere how we determine which tensors are compressed vs uncompressed (rank 1 & the compression threshold?)

SciPioneer replied:
It's specified in `_should_compress` method, noted in the precursor PR.

rohan-varma reviewed (approved):
LGTM! Thanks for refactoring this
